mock unit test for source sentry

suisuixia42 opened the pull request:
## What
Added comprehensive unit tests for the source-sentry connector using mock server testing pattern.


## User Impact
<!--
* What is the end result perceived by the user?
* If there are negative side effects, please list them. 
-->

## Can this PR be safely reverted and rolled back?
<!--
* If unsure, leave it blank.
-->
- [ ] YES 💚
- [ ] NO ❌

suisuixia42 committed:
mock unit test for source sentry

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/airbytehq/airbyte?pullRequest=70234) <br/>All committers have signed the CLA.

octavia-squidington-iii committed:
chore: auto-fix lint and format issues

aaronsteers committed:
Merge branch 'master' into unit-test-for-source-sentry-connector

aaronsteers committed:
Merge branch 'master' into unit-test-for-source-sentry-connector

aaronsteers committed:
Merge branch 'master' into unit-test-for-source-sentry-connector

brianjlai commented on airbyte-integrations/connectors/source-sentry/unit_tests/integration/test_issues.py:
```diff
@@ -0,0 +1,63 @@
+        catalog = CatalogBuilder().with_stream(_STREAM_NAME, SyncMode.full_refresh).build()
+        output = read(source, config=self._config(), catalog=catalog)
+
+        assert len(output.records) == 2, f"Expected 2 issues from 2 pages"
```
we should assert on some of the values not just the number of records like at least the `id` of the expected records

brianjlai commented on airbyte-integrations/connectors/source-sentry/unit_tests/integration/test_events.py:
```diff
@@ -0,0 +1,119 @@
+        # ARRANGE
+        http_mocker.get(
+            SentryRequestBuilder.events_endpoint(_ORGANIZATION, _PROJECT, _AUTH_TOKEN).build(), create_response("events", has_next=False)
+        )
```
If you look at the implementation of the `events` stream, we also include a query parameter for `full=true`, we should also be including this in the mock request we expect to validate we're passing that in properly.

I don't see that in the expected request unless I'm mistaken

brianjlai commented on airbyte-integrations/connectors/source-sentry/unit_tests/integration/test_events.py:
```diff
@@ -0,0 +1,119 @@
+
+
+@freezegun.freeze_time(_NOW.isoformat())
+class TestEventsStream(TestCase):
```
The `events` stream is an incremental sync that allows for passing the date range as query parameters in the API request.

We currently have a test case `test_full_refresh_single_page` which validates the first sync attempt in which we have no state message. But we should also be modifying that test to validate that `start` and `end` query parameters are included in the outbound request. We should also validate the state message we receive back

We also should add a second test where we pass in state with a later date  and validate that the later date is used in the outbound API request. We should also validate the state message we receive back

brianjlai reviewed (commented):
I left a few comments related to getting better coverage of this connector.

The main things that we should try to also capture in the test are:

- incremental syncs where we pass in query parameters in the outbound API request and therefore should mock to include them
- data feed incremental syncs where we validate that older records than the current state are not included
- more pagination coverage

Otherwise this is shaping up pretty well so far!

suisuixia42 committed:
adding more tests

suisuixia42 committed:
Merge remote-tracking branch 'origin/unit-test-for-source-sentry-connector' into unit-test-for-source-sentry-connector

# Conflicts:
#	airbyte-integrations/connectors/source-sentry/unit_tests/integration/request_builder.py
#	airbyte-integrations/connectors/source-sentry/unit_tests/integration/test_events.py
#	airbyte-integrations/connectors/source-sentry/unit_tests/integration/test_issues.py
#	airbyte-integrations/connectors/source-sentry/unit_tests/integration/test_project_detail.py
#	airbyte-integrations/connectors/source-sentry/unit_tests/integration/test_projects.py
#	airbyte-integrations/connectors/source-sentry/unit_tests/integration/test_releases.py

suisuixia42 committed:
fix: Restore all integration tests after merge overwrote changes

- Restored test_issues.py (305 lines): all 6 tests including smart stopping
- Restored test_releases.py (218 lines): all 5 tests including smart stopping
- Restored test_projects.py (181 lines): all 4 tests including client-side filtering
- Restored test_events.py (212 lines): all 4 tests including URL param validation
- Added with_query_params() method to request_builder.py for Events tests

All 20 integration tests passing ✅

octavia-squidington-iii committed:
chore: auto-fix lint and format issues

brianjlai commented on airbyte-integrations/connectors/source-sentry/unit_tests/integration/test_issues.py:
```diff
@@ -0,0 +1,300 @@
+        assert state["lastSeen"].startswith("2024-01-20T15:00:00"), f"Expected state cursor to be latest record's lastSeen, got {state}"
+
+    @HttpMocker()
+    def test_incremental_pagination_with_smart_stopping(self, http_mocker: HttpMocker):
```
nit: I don't really like this name `with_smart_stopping`. It feels very vague and ambiguous. We should call this `with_data_feed` or something like that because that's the CDK/sentry feature that we're actually testing

brianjlai reviewed (approved):
:shipit: changes look good to me! The main nit I have is that we should avoid calling something smart_stopping because it isn't a good description of the feature we're testing which is really the `is_data_feed`.

And just an fyi, I adjusted our CI a little bit so that hopefully changes to the `/integration` and `/unit_tests` folders shouldn't require a version bump which is the current CI check failing. After you push any new commits to the PR, let's double check the results and validate that the error goes away. If it doesn't we can just force merge or bump the patch, but ideally that CI step should get skipped

suisuixia42 committed:
motify the name for tests
